### PR TITLE
"merge" subcommand

### LINKED ISSUE #105: Merge multiple non overlapping cosmogony files
To generate a cosmogony of the world, you currently need a lot of ram since the whole planet.osm.pbf needs to be loaded.

We could read split non overlapping osm files, without a shared parent (eg. split by continent or country) and merge the generated cosmogony.

Since there is no contract on the indexes (`id` and `parent`) being consecutive, it seems a dumb offset by files, so each files will have ids in a separate range, will do the job.

I have a first shot at this, and need to test it on real cases but do you have any thoughts on this (I haven't worked on cosmogony for a while, maybe I'm completely off the mark)?

ping @prhod @amatissart

### PR DESCRIPTION
**Pull request by antoine-de**
fixes #105 

to run it:

in cosmogony_merger directory:

`cargo run -- *.jsonl -o merged_cosmo.jsonl`

Not sure about this, I tested it on the samples files and it seems ok, but I cannot test this more thoroughly now. Can you test @prhod ?

I can put this in another repository if you feel this does not belongs here. (and will add some tests/doc if I can :stuck_out_tongue_winking_eye: )

### CODE REVIEW

**Comment by prhod** [comment]
Thank you so much :pray: 
I don't have expertise on Rust project structure, and I'd be comfortable at the end to have a Docker image containing both the converter and merger. Give me some time for testing it please :wink:

**Comment by antoine-de** [comment]
I can make a dokerfile with it if you want. but next week, I'm off for the week end 😉

File: cosmogony/Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,633 @@
+# This file is automatically @generated by Cargo.
```
Comment (amatissart) [inline cosmogony/Cargo.lock:1]: This `cosmogony/Cargo.lock` is not supposed to be committed, right ? As `cosmgony/` is used to publish the `cosmogony` crate.

**Comment by amatissart** [comment]
It looks good to me.  
From a usability and maintenance perspective, I wonder if this feature could not be included in the existing Docker image. 

Maybe this could even be defined as a [subcommand](https://docs.rs/structopt/0.3.20/structopt/#subcommands) in the existing executable, to be able to run:
```
cosmogony merge *.jsonl -o merged_cosmo.jsonl
```
What do you think ?

**Comment by prhod** [comment]
a subcommand could be nice! :+1:

**Comment by prhod** [comment]
I've merged europe and africa on http://cosmogony.world and it works great :+1:

**Comment by antoine-de** [comment]
the subcommand is a great idea!
Thus I can put the merge part in the binary and remove the cosmogony_merger sub directory (and the docker would have it for free).

For the sake of consistency, maybe I can also add a `cosmogony generate` subcommand (and keep the old way of working not to break integrations). Would that be ok for you?

**Comment by amatissart** [comment]
> For the sake of consistency, maybe I can also add a `cosmogony generate` subcommand (and keep the old way of working not to break integrations). Would that be ok for you?

It sounds good to me :+1:

**Comment by prhod** [comment]
+1 nice idea :smile:

**Comment by antoine-de** [comment]
oups sorry I forgot this :see_no_evil: 

I added the merge as subcommand and kept the retrocompatibility. I did not find a nice way to handle this with structopt/clap, so I did it a bit manually.

File: cosmogony/Cargo.lock (hunk 1 above)
Comment (amatissart) [reply cosmogony/Cargo.lock:1]: My question still stands.

File: src/bin/cosmogony.rs [hunk 2]
```diff
@@ -38,6 +67,24 @@ Accepted extensions are '.json', '.json.gz', '.jsonl', '.jsonl.gz'
...
+#[derive(StructOpt, Debug)]
+struct MergeArgs {
+    /// Cosmogony files to process
+    #[structopt(name = "FILE", parse(from_os_str))]
+    files: Vec<PathBuf>,
+    /// output file name
+    #[structopt(
+        short = "o",
+        long = "output",
+        default_value = "cosmogony.json",
```
Comment (amatissart) [inline src/bin/cosmogony.rs:79]: This default value causes panic, as `.json` format is not supported as the output format. It should be removed or replaced with a supported extension.

File: src/bin/cosmogony.rs [hunk 3]
```diff
@@ -112,8 +166,12 @@ fn init_logger() {
 
 fn main() {
     init_logger();
-    let args = Args::from_args();
-    if let Err(e) = cosmogony(args) {
+    // Note: for retrocompatibility, we also try to read the args without subcommand
+    // to generate a cosmogony
+    let args = GenerateArgs::from_args_safe().map(|a| Args::Generate(a)).unwrap_or_else(|_| {
+        Args::from_args()
+    });
```
Comment (amatissart) [inline src/bin/cosmogony.rs:173]: `cargo fmt` complains here

File: src/merger.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+pub fn merge_cosmogony(files: &[PathBuf], output: &Path) -> Result<(), failure::Error> {
+    let mut merger = CosmogonyMerger::default();
+
+    let format = OutputFormat::from_filename(output)?;
+    let file = std::fs::File::create(output)?;
+    let mut stream = std::io::BufWriter::new(file);
+    match format {
+        OutputFormat::JsonGz | OutputFormat::Json => panic!(
+            "cannot read real cosmogonies, only jsonl/jsonl.gz to be able to stream the files"
```
Comment (amatissart) [inline src/merger.rs:70]: This error message does not sound relevant. The _output_ format is validated here, not the files that will be read.  
As far as I can see, nothing prevents from using .json or .json.gz formats as input `files`.

**Review by amatissart — COMMENTED** [review]
Thanks for this. I made a few comments, but it looks good to me overall.

File: cosmogony/Cargo.lock (hunk 1 above)
Comment (antoine-de) [reply cosmogony/Cargo.lock:1]: yep, sorry, completly forgot this :/

File: src/merger.rs (hunk 4 above)
Comment (antoine-de) [reply src/merger.rs:70]: hum I only added the method to write streams (`to_json_stream`) as we write the zones on the go (and I don't see a usecase where someone would want to output `json` here :thinking: 

would that be okay for you if I only correct the message?

**Review by amatissart — APPROVED** [review]
@antoine-de Feel free to merge this PR if you don't wait for another validation.

**Comment by antoine-de** [comment]
thanks for the review!

**Pull request merged**